exchanges/exchanges.go

```diff
@@ -2851,3 +2887,285 @@ func (poloniex *PoloniexExchange) Refresh() {
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("epoch_order Unmarshal error: %v", err))
+				return
+			}
+			dcr.checkSeq(note.Seq)
+			return // Skip wsUpdate. Nothing has changed.
+		}
+	}
+	dcr.wsUpdated()
+}
+
+// checkSeq verifies that the seq is sequential, and increments the seq counter.
+func (dcr *DecredDEX) checkSeq(seq uint64) bool {
+	if seq != dcr.seq+1 {
+		dcr.setWsFail(fmt.Errorf("incorrect sequence. wanted %d, got %d", dcr.seq+1, seq))
+		return false
+	}
+	dcr.seq = seq
+	return true
+}
+
+// lastStamp is the unix timestamp of the received response or notification.
+func (dcr *DecredDEX) lastStamp() int64 {
+	dcr.orderMtx.RLock()
+	defer dcr.orderMtx.RUnlock()
+	return dcr.stamp
+}
+
+// setOrderBook processes the order book data from 'orderbook' request.
+// setOrderBook should only be called with the orderMtx write-locked.
+func (dcr *DecredDEX) setOrderBook(ob *msgjson.OrderBook) {
+	dcr.buys = make(wsOrders)
+	dcr.asks = make(wsOrders)
+	dcr.ords = make(map[string]*msgjson.BookOrderNote)
+	dcr.seq = ob.Seq
+
+	addToSide := func(side wsOrders, ord *msgjson.BookOrderNote) {
+		bucket := side.order(int64(ord.Rate), float64(ord.Rate)/1e8)
+		bucket.volume += float64(ord.Quantity) / 1e8
+		dcr.ords[ord.OrderID.String()] = ord
+	}
+
+	for _, ord := range ob.Orders {
+		if ord.Side == msgjson.BuyOrderNum {
+			addToSide(dcr.buys, ord)
+		} else {
+			addToSide(dcr.asks, ord)
+		}
+	}
+	dcr.wsInitialized()
+
+	depth := dcr.wsDepthSnapshot()
+
+	dcr.Update(&ExchangeState{
+		Price: depth.MidGap(),
+		// Change:       priceChange, // With candlesticks
+		Stamp: dcr.stamp,
+		// Candlesticks: candlesticks, // Not yet
+		Depth: depth,
+	})
+}
+
+// bookOrder processes the 'book_order' notification.
+// bookOrder should only be called with the orderMtx write-locked.
+func (dcr *DecredDEX) bookOrder(ord *msgjson.BookOrderNote) {
+	side := dcr.asks
+	if ord.Side == msgjson.BuyOrderNum {
+		side = dcr.buys
+	}
+	bucket := side.order(int64(ord.Rate), float64(ord.Rate)/1e8)
+	bucket.volume += float64(ord.Quantity) / 1e8
+	dcr.ords[ord.OrderID.String()] = ord
+}
+
+// unbookOrder processes the 'unbook_order' notification.
+// unbookOrder should only be called with the orderMtx write-locked.
+func (dcr *DecredDEX) unbookOrder(note *msgjson.UnbookOrderNote) {
+	oid := note.OrderID.String()
+	ord := dcr.ords[oid]
+	delete(dcr.ords, oid)
+	if ord == nil {
+		dcr.setWsFail(fmt.Errorf("no order found to unbook"))
+		return
+	}
+	side := dcr.asks
+	if ord.Side == msgjson.BuyOrderNum {
+		side = dcr.buys
+	}
+	rateKey := int64(ord.Rate)
+	bucket := side.order(rateKey, float64(ord.Rate)/1e8)
+	bucket.volume -= float64(ord.Quantity) / 1e8
+	if bucket.volume < 1e-8 {
+		delete(side, rateKey)
+	}
+}
+
+// updateRemaining processes the 'update_remaining' notification.
+// updateRemaining should only be called with the orderMtx write-locked.
+func (dcr *DecredDEX) updateRemaining(update *msgjson.UpdateRemainingNote) {
+	oid := update.OrderID.String()
```

Comment:
I think OrderID could be nil if the server made it so with null json.  Just want to be defensive.  Although I think (Bytes).String might actually not be a problem even with nil slice